File: internal/sandbox/runtimes/python/entrypoint.py
#!/usr/bin/env python3
"""
Sandbox entrypoint for Python execution.

Protocol:
    stdin  → JSON { code, input, context, params }
    stdout → JSON output (value of `output` variable after execution)
    stderr → logs / errors

The user's code runs in a restricted namespace. Available globals:
    input, context, params, print (→ stderr)

Call `output(val)` to produce a result.
"""

import json
import os
import sys
import traceback


def main():
    raw = sys.stdin.read()
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError as e:
        print(f"entrypoint: invalid JSON payload: {e}", file=sys.stderr)
        sys.exit(1)

    code = payload.get("code", "")
    input_data = payload.get("input")
    context = payload.get("context", {})
    params = payload.get("params", {})

    # Redirect print to stderr so stdout stays clean for output JSON
    original_print = print

    def sandbox_print(*args, **kwargs):
        kwargs["file"] = sys.stderr
        original_print(*args, **kwargs)

    # Unified output() function — like Lambda handler pattern
    _output_value = None
    _output_set = False

    def output_fn(val):
        nonlocal _output_value, _output_set
        _output_value = val
        _output_set = True

    # Build execution namespace
    namespace = {
        "input": input_data,
        "context": context,
        "params": params,
        "output": output_fn,
        "print": sandbox_print,
        # Standard safe builtins
        "len": len,
        "range": range,
        "enumerate": enumerate,
        "zip": zip,
        "map": map,
        "filter": filter,
        "sorted": sorted,
        "reversed": reversed,
        "list": list,
        "dict": dict,
        "set": set,
        "tuple": tuple,
        "str": str,
        "int": int,
        "float": float,
        "bool": bool,
        "type": type,
        "isinstance": isinstance,
        "hasattr": hasattr,
        "getattr": getattr,
        "setattr": setattr,
        "min": min,
        "max": max,
        "sum": sum,
        "abs": abs,
        "round": round,
        "any": any,
        "all": all,
        "json": json,
        "True": True,
        "False": False,
        "None": None,
        "__builtins__": {},
    }

    # Debug mode: use debugpy for DAP-based interactive debugging
    if os.environ.get("SANDBOX_DEBUG"):
        script_path = "/sandbox/user_script.py"
        with open(script_path, "w") as f:
            f.write(code)

        try:
            import debugpy
        except ImportError:
            print("debugpy not installed — use debug image", file=sys.stderr)
            sys.exit(1)

        debugpy.listen(("0.0.0.0", 5678))
        print("debugpy: waiting for client on port 5678...", file=sys.stderr)
        debugpy.wait_for_client()
        print("debugpy: client attached", file=sys.stderr)

        try:
            exec(compile(code, "user_script.py", "exec"), namespace)
        except Exception:
            traceback.print_exc(file=sys.stderr)
            sys.exit(1)
    else:
        try:
            exec(compile(code, "sandbox.py", "exec"), namespace)
        except Exception:
            traceback.print_exc(file=sys.stderr)
            sys.exit(1)

    result = _output_value if _output_set else None

    try:
        sys.stdout.write(json.dumps(result))
    except (TypeError, ValueError) as e:
        print(f"output serialization error: {e}", file=sys.stderr)
        sys.stdout.write(json.dumps(str(result)))

    sys.exit(0)


if __name__ == "__main__":
    main()

File: internal/sandbox/runtimes/php/entrypoint.php
#!/usr/bin/env php
<?php
/**
 * Sandbox entrypoint for PHP execution.
 *
 * Protocol:
 *   stdin  → JSON { code, input, run_input, context, config }
 *   stdout → JSON output (from output() call)
 *   stderr → logs / errors
 *
 * User code runs with $input, $run_input, $context, $config available.
 * Call output($val) to produce a result.
 * echo/print → stderr (stdout reserved for JSON output).
 */

$raw = file_get_contents('php://stdin');
$payload = json_decode($raw, true);

if ($payload === null && json_last_error() !== JSON_ERROR_NONE) {
    fwrite(STDERR, "entrypoint: invalid JSON payload: " . json_last_error_msg() . "\n");
    exit(1);
}

$code      = $payload['code'] ?? '';
$input     = $payload['input'] ?? null;
$run_input = $payload['run_input'] ?? null;
$context   = $payload['context'] ?? [];
$config    = $payload['config'] ?? [];

// output() — sole output mechanism
$_output_value = null;
$_output_set = false;

function output($val) {
    global $_output_value, $_output_set;
    $_output_value = $val;
    $_output_set = true;
}

// Redirect echo/print to stderr via output buffering
ob_start(function ($buffer) {
    fwrite(STDERR, $buffer);
    return '';
}, 1);

try {
    eval($code);
} catch (Throwable $e) {
    fwrite(STDERR, "runtime error: " . $e->getMessage() . "\n");
    fwrite(STDERR, $e->getTraceAsString() . "\n");
    ob_end_clean();
    exit(1);
}

ob_end_clean();

$result = $_output_set ? $_output_value : null;

$json = json_encode($result);
if ($json === false) {
    fwrite(STDERR, "output serialization error: " . json_last_error_msg() . "\n");
    $json = json_encode(strval($result));
}

fwrite(STDOUT, $json);
exit(0);
